cmd/dra-example-controller/allocations.go

```diff
@@ -0,0 +1,112 @@
+/*
+ * Copyright 2023 The Kubernetes Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package main
+
+import (
+	"sync"
+
+	nascrd "github.com/kubernetes-sigs/dra-example-driver/api/example.com/resource/gpu/nas/v1alpha1"
+)
+
+type PerNodeAllocatedClaims struct {
+	sync.RWMutex
+	allocations map[string]map[string]nascrd.AllocatedDevices
+}
+
+func NewPerNodeAllocatedClaims() *PerNodeAllocatedClaims {
+	return &PerNodeAllocatedClaims{
+		allocations: make(map[string]map[string]nascrd.AllocatedDevices),
+	}
+}
+
+func (p *PerNodeAllocatedClaims) Exists(claimUID, node string) bool {
+	p.RLock()
+	defer p.RUnlock()
+
+	_, exists := p.allocations[claimUID]
+	if !exists {
+		return false
+	}
+
+	_, exists = p.allocations[claimUID][node]
+	return exists
+}
+
+func (p *PerNodeAllocatedClaims) Get(claimUID, node string) nascrd.AllocatedDevices {
+	p.RLock()
+	defer p.RUnlock()
+
+	if !p.Exists(claimUID, node) {
+		return nascrd.AllocatedDevices{}
+	}
+	return p.allocations[claimUID][node]
+}
+
+func (p *PerNodeAllocatedClaims) VisitNode(node string, visitor func(claimUID string, allocation nascrd.AllocatedDevices)) {
+	p.RLock()
+	for claimUID := range p.allocations {
+		if allocation, exists := p.allocations[claimUID][node]; exists {
+			p.RUnlock()
```

Comment:
allocation is a copy. Why do we need to unlock?

Reply:
We only need the lock to avoid a concurrent map access on `p.allocations[claimUID][node]`. Once we have pulled `allocation` out of it, we unlock, call the visit() call back with it and then relock once the `visit()` callback has completed, so that we hold the lock for the next iteration of the loop.

Reply:
ok, thanks for the explanations.